tools/qlog/src/lib.rs

```diff
@@ -258,6 +298,264 @@ impl Default for Qlog {
+
+                    Some(TimeUnits::Us) => event_time.as_micros().to_string(),
+
+                    None => String::from(""),
+                },
+
+                None => String::from(""),
+            };
+
+            // TODO: this unfortunately removes any frames already existing in the
+            // event, which might be undesirable.
+            let (ev_data, contains_frames) =
+                match serde_json::to_string(&event.data) {
+                    Ok(mut ev_data_out) =>
+                        if event.data.contains_quic_frames() {
+                            if let Some(index) = ev_data_out.find("\"frames\":[")
+                            {
+                                ev_data_out
+                                    .truncate(index + "\"frames\":[".len());
+                            }
+
+                            self.first_frame = true;
+
+                            (ev_data_out, true)
+                        } else {
+                            (ev_data_out, false)
+                        },
+
+                    _ => return Err(Error::Done),
+                };
+
+            let maybe_comma = if self.first_event {
+                self.first_event = false;
+                ""
+            } else {
+                ","
+            };
+
+            let maybe_terminate = if contains_frames { "" } else { "]" };
+
+            let ev_time =
+                serde_json::to_string(&EventField::RelativeTime(rel)).ok();
+            let ev_cat =
+                serde_json::to_string(&EventField::Category(event.category)).ok();
+            let ev_ty = serde_json::to_string(&EventField::Event(event.ty)).ok();
+
+            if let (Some(ev_time), Some(ev_cat), Some(ev_ty)) =
+                (ev_time, ev_cat, ev_ty)
+            {
+                let out = format!(
+                    "{}[{},{},{},{}{}",
+                    maybe_comma, ev_time, ev_cat, ev_ty, ev_data, maybe_terminate
+                );
+
+                self.writer.as_mut().write_all(out.as_bytes())?;
+                self.writing_event = contains_frames;
+
+                return Ok(contains_frames);
+            } else {
+                return Err(Error::Done);
+            }
+        }
+
+        Ok(false)
+    }
+
+    /// Writes a JSON-serialized `QuicFrame`.
+    ///
+    /// Only valid while in the frame-serialization mode.
+    pub fn write_frame(&mut self, frame: QuicFrame, last: bool) -> Result<()> {
+        if !self.writing_event || self.log_finished {
+            return Err(Error::InvalidState);
+        }
+
+        match serde_json::to_string(&frame) {
+            Ok(mut out) => {
+                if !self.first_frame {
+                    out.insert(0, ',');
+                } else {
+                    self.first_frame = false;
+                }
+
+                self.writing_frames = true;
+
+                self.writer.as_mut().write_all(out.as_bytes())?;
+
+                if last {
```

Comment:
Looking at the quiche code, this is never set to `true` so do we really need it?

Reply:
Though if needed, should maybe just call `finish_frames()` to avoid duplicating the code.

Reply:
added finish_frames, to ensure state transitions.